neuro_san/internals/tool_factories/service_tool_factory_provider.py

```diff
@@ -58,14 +73,32 @@ def add_agent_tool_registry(self, agent_name: str, registry: AgentToolFactory):
         """
         Register existing agent tool registry
         """
+        is_new: bool = False
         with self.lock:
-            is_new: bool = self.agents_table.get(agent_name, None) is None
+            is_new = self.agents_table.get(agent_name, None) is None
             self.agents_table[agent_name] = registry
+        # Notify listeners about this state change:
+        # do it outside of internal lock
+        for listener in self.listeners:
             if is_new:
-                self.logger.info("BUILT tool registry for agent %s", agent_name)
+                listener.agent_added(agent_name)
+                self.logger.info("ADDED tool registry for agent %s", agent_name)
             else:
+                listener.agent_modified(agent_name)
                 self.logger.info("REPLACED tool registry for agent %s", agent_name)
```

Comment:
Depending on whether agent registry being added is really new or replacing existing registry item in ServiceToolFactory, call appropriate listener method.

Reply:
Is it really the case that if it's not is_new it's always modified?
Maybe a clarifying comment in add_agent_tool_registry() method comment as to when it is expected to be called.  My prepping on current code (not this PR) shows it's called from AgentServer and DirectAgentSessionFactory, but maybe there is a new way in given the new machinery?

Reply:
This is actually kinda interesting. 
We have been thinking about being able to specify a few other factories via the agent hocon file (like LlmFactory and BaseToolFactory) to allow each agent to have their own definitions instead of a server-global definition.  This  could make *those* config files re-read when agents are updated too which I think would fall in line with expectations.  That is, right now those global guys are likely not on the filesystem-change radar.

Reply:
> Is it really the case that if it's not is_new it's always modified?
   That's a caller responsibility to only call this method for existing agent if its definition (AgentToolFactory) has really changed. FWIW though, call with unchanged AgentToolFactory is actually a no-op right now.